server.js

```diff
@@ -14,12 +14,68 @@
  * limitations under the License.
  */
 
-const express = require("express");
+const express = require('express');
+const locale = require('locale');
+const path = require('path');
+const fs = require('fs');
+const cookieParser = require('cookie-parser');
+
 const app = express();
 
-app.use(express.static("dist"));
-// app.use(express.static("dist/en"));
-app.use(express.static("dist/ja"));
+const staticFilesPath = [
+  'images', 'test', 'app.css', 'app.js'
+];
+
+// Detect all locale directories
+const supportedLocales = [];
+const dirs = fs.readdirSync(path.join(__dirname, 'dist'));
+for (const dir of dirs) {
+  if (staticFilesPath.includes(dir)) continue;
+  supportedLocales.push(dir);
+}
+const defaultLocale = 'en';
+
+app.use(locale(supportedLocales, defaultLocale));
+app.use(cookieParser());
+app.use(express.static('dist'));
+
+app.get("*", function(req, res) {
+  const dir = req.path.split(path.sep)[1];
+  // Serve static files as is
+  if (staticFilesPath.includes(dir)) {
+    const filePath = path.join(__dirname, 'dist', req.path);
+    res.sendFile(filePath);
+    return;
+  }
+
+  let lang = req.query.hl;
+  if (!lang || !supportedLocales.includes(lang)) {
+    // `hl` is NOT specified.
+    if (req.cookies.preferred_lang) {
+      lang = req.cookies.preferred_lang;
+    } else if (req.locale) {
+      lang = req.locale;
+    } else {
+      lang = defaultLocale;
+    }
+  }
+  // Update preferred language (with sanity check)
+  if (req.query.preferred_lang !== undefined &&
+      supportedLocales.includes(lang)) {
+    res.cookie('preferred_lang', lang,
+      {maxAge: 10 * 365 * 24 * 60 * 60 * 1000, httpOnly: true});
+  }
+
+  // Serve the lang specific file from a localized directory
+  const localizedFilePath = path.join(__dirname, 'dist', lang, req.path, 'index.html');
```

Comment:
This is a security hole. I could set my `preferred_lang` cookie to be "../../../escape/the/root".

Can we check that the lang is in "[-_a-zA-Z]" (i.e., '-', '_' or English upper/lower case)?

Reply:
GitHub ate my underscore and treated it as italics. It should be, any underscore, dash, or English upper/lower case.